keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,360 @@
+
+Delegating EndpointSlice control would allow custom controllers to define their own criteria for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As a reference implementation, and since [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.
+
+### Goals
+
+* Provide the ability to disable the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for particular services.
+
+### Non-Goals
+
+* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers.
+* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
+* Modify the Service / EndpointSlice / Endpoints Specs.
+* Add new features to the EndpointSlice Mirroring Controller.
+
+## Proposal
+
+#### Well-Known Label
+
+`service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, no matter the service specs, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers for that service will be disabled, thus Endpoints and EndpointSlices for this service will not be created by the Kubernetes Controller Manager. If the label is not set, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers will be enabled for that service and the Endpoints and EndpointSlices will be handled as of today.
+
+The Kubernetes Controller Manager will implement this label both at object creation and on dynamic addition/removal/updates of this label.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)).
+
+Therefore, it must be possible to disable the default Kubernetes Endpoints and EndpointSlice Controller for certain services and use a specialized EndpointSlice reconciler implementation to create a controller for secondary network providers.
```

Comment:
It's not clear where you're using "service" to mean `v1.Service` and where you're using it in a generic sense...

How is the Service configured? Is it `type: ClusterIP`? Does it have the `service-proxy-name` label set?

Who is calling the Service and from where? (Pods that share the same secondary network? Pods that _don't_ share the same secondary network? Nodes? Cluster-external hosts?) What is the expectation for if the "wrong" people try to connect to it?

(You have both "for example, [Multus]" and "e.g. [Multus]" there btw. Second one can probably go away.)

Reply:
I always meant the Kubernetes service, I added it for clarity, thank you.
I also added the mention to the service-proxy label. But I think the rest of the configuration of the service doesn't really matter, it could be both internal service traffic or traffic originating from outside of the cluster. I am not sure we should try to fully describe secondary networks and services here.

Reply:
> But I think the rest of the configuration of the service doesn't really matter

Ideally, user stories should help the reader understand why you made some decisions and not others. But they can only do that if the user stories are sufficiently clear and evocative.

(And you _especially_ don't want your user stories to be making the reviewers think about things like "but how will `type: LoadBalancer` services work if the load balancer doesn't have access to the secondary network?")